### DISCUSSION TITLE
Migration path for "object is being merged into a Session along the backref cascade path for relationship"

### DISCUSSION THREAD

**Discussion by ianjosephwilson** [sqlalchemy/sqlalchemy#7693] [Usage Questions] [answered]
I really enjoy sqlalchemy and I'm looking forward to 2.0.  I just turned on `SQLALCHEMY_WARN_20` for my code base and I have a lot of `object is being merged into a Session along the backref cascade path for relationship`.  I thought I had this warning figured out but I'm still a little confused about why/when/how this is reported and also what the desired usage would look like.

Assuming I have 2 model classes with a typical 1toMany relationship as follows:

```python
class Comment(Base):
    __tablename__ = "comments"
    id = Column(Integer, primary_key=True, autoincrement=True)
    discussion_id = Column(Integer, ForeignKey('discussions.id'), nullable=False)


class Discussion(Base):
    __tablename__ = "discussions"
    id = Column(Integer, primary_key=True, autoincrement=True)
    comments = relationship(Comment, backref="discussion")


with Session(engine) as session:
    with session.begin():
        
        d = Discussion()
        session.add(d)
        # Warning is raised here
        c = Comment(discussion=d)
        # But we add c directly after creation
        session.add(c)
```

This code seems like it would work fine in either case, ie. `cascade_backrefs=True` or `cascade_backrefs=False`.  The issue I guess is if I didn't include `session.add(c)` and was using 2.0 I would not realize that the comment would not be created.  Additionally if I was to take this code into 2.0 WITHOUT `session.add(c)` then I'd have a bug.  As I understand things that is why the warning exists.  But in my case I'm trying to prepare 1.4 code for 2.0.   So I guess my questions are as follows:

1. What is the best way to sift through these specific warnings in a systematic way?  I can't seem to easily rewrite the code to avoid the warning and if I turn off cascade_backrefs for a relationship there is a chance I didn't address each one and globally turning it off would be a bad idea with `future=True`.  Essentially this warning is more of a "There might be a mistake here" so I can't set the warning to an error because then I can't go through each warning.
2. Is there a more idiomatic 2.0 way of writing this code?  Ie. a way to avoid this warning but still let different parts of code create and link different objects.  Is a better solution to not use backrefs and just explicitly set up both sides of each relationship ?
3. In 2.0 this seems like it will be a bug factory for me.  I can't tell if its because I'm so used to the safety net of 1.X series or not but I think I will forget to add some deep nested object on a regular basis.  Will there be a way to detect this mistake other than my own  tests?  Ie. I will want this same warning in 2.0 but it won't be available because `cascade_backrefs` will cease to exist.  Ie. Warning: object linked in relationship backref not saved because it is not added in the session.

**Comment by zzzeek** [comment, marked as answer]
>     1. What is the best way to sift through these specific warnings in a systematic way?  I can't seem to easily rewrite the code to avoid the warning and if I turn off cascade_backrefs for a relationship there is a chance I didn't address each one and globally turning it off would be a bad idea with `future=True`.  Essentially this warning is more of a "There might be a mistake here" so I can't set the warning to an error because then I can't go through each warning.
> 
>     2. Is there a more idiomatic 2.0 way of writing this code?  Ie. a way to avoid this warning but still let different parts of code create and link different objects.  Is a better solution to not use backrefs and just explicitly set up both sides of each relationship ?

you should definitely use backrefs, they are kind of necessary when you have complementary relationships.  right now, the warning that's emitted as you've observed can be turned on or off at the level of a particular class-level relationship.  but here it seems you are interested in alleviating the warning on a case by case basis within the scope of a particular relationship where you don't want to change them all at once.  I suppose if this is a large codebase then maybe that's a requirement.   

you can rewrite object interactions to avoid the warning by making sure both of the two objects in question are either not in the session, or are in the session, when they are associated with each other.    This might not be feasible depending on how your code is structured, but to the degree you can re-order operations in this way, that would eliminate the warning on a case by case basis.     

given:

```py
d = Discussion()
session.add(d)
c = Comment(discussion=d)
session.add(c)
```

this can be written either as:

```py
d = Discussion()
c = Comment(discussion=d)
session.add(d)
session.add(c)  # optional, d.cs would normally cascade this
```

or as:

```py
d = Discussion()
session.add(d)
c = Comment()
session.add(c)
c.discussion = d
```

if you need more comprehensive techniques for object configurations that are not readily re-arrangeable, and it's not an option to just fix all the warnings for a particular relationship at once, I propose a short linter below that will allow you to turn off cascade_backrefs for the relationship / globally first and the linter will alert for any objects that weren't attached by commit time.


> 
>     3. In 2.0 this seems like it will be a bug factory for me.  I can't tell if its because I'm so used to the safety net of 1.X series or not but I think I will forget to add some deep nested object on a regular basis.

The behavior change here doesn't disable cascading of nested objects, if you have an A() that is present in the Session and you then attach A->B, B->C, C->D etc. those all get added.  The thing that's disabled is cascading when you create the association from the B->A side.    We've had lots of users for years being surprised by this behavior, running into issues where they want to set some B->A and not have B get immediately yanked into the session, usually B would not be in a state that's ready for flushing and they'd have autoflush errors.

> Will there be a way to detect this mistake other than my own  tests? 

I propose a short recipe below.  However, put this in perspective - if you create an A, and dont session.add() it, it wont get persisted - so the mistake of not adding an object is already one that you need to have your own tests for.   I dont think this is so different than that, make an object, it has to be add()ed or associated on the right-hand side of another already-associated object.

> Ie. I will want this same warning in 2.0 but it won't be available because `cascade_backrefs` will cease to exist.  Ie. Warning: object linked in relationship backref not saved because it is not added in the session.

yes you use events to track this,   i think saving up all the objects that you expect to be attached and then checking for that list to be empty at commit time is sufficient.   See the example below


```py
from sqlalchemy import event
from sqlalchemy.orm import object_session
from sqlalchemy.orm import RelationshipProperty


def configure_listener(class_, key, inst):
    """an 'attribute_instrument' listener.

    this is from the example at
    https://docs.sqlalchemy.org/en/14/_modules/examples/custom_attributes/listen_for_events.html

    """
    if not isinstance(inst.property, RelationshipProperty):
        return

    @event.listens_for(inst, "append")
    def append(instance, value, initiator):
        _set_needs_attach(instance, value)

    @event.listens_for(inst, "set")
    def set_(instance, value, oldvalue, initiator):
        _set_needs_attach(instance, value)


    def _set_needs_attach(instance, value):
        """act on "instance.foo = value"

        if "value" is in a session and "instance" isn't, add it to a special
        collection in that session

        """

        sess = object_session(value)
        if sess and not object_session(instance):
            if "needs_attach" not in sess.info:
                sess.info["needs_attach"] = set()
            sess.info["needs_attach"].add(instance)


def update_attached(session, instance):
    """operate on the after_attach event.  remove the instance from
    needs_attached"""

    if "needs_attach" in session.info:
        session.info["needs_attach"].discard(instance)


def verify_attached(session):
    """operate on the before_commit event.   everything should be attached."""

    if "needs_attach" in session.info:
        assert not session.info[
            "needs_attach"
        ], f"objects were unattached : {session.info['needs_attach']}"


if __name__ == "__main__":

    from sqlalchemy import Column, Integer, String, ForeignKey, create_engine
    from sqlalchemy.orm import relationship
    from sqlalchemy.orm import declarative_base
    from sqlalchemy.orm import sessionmaker

    Base = declarative_base()

    e = create_engine("sqlite://")
    smaker = sessionmaker(e, future=True)

    event.listen(Base, "attribute_instrument", configure_listener)
    event.listen(smaker, "after_attach", update_attached)
    event.listen(smaker, "before_commit", verify_attached)

    class A(Base):
        __tablename__ = "a"

        id = Column(Integer, primary_key=True)
        data = Column(String)
        bs = relationship("B", back_populates="a")

    class B(Base):
        __tablename__ = "b"
        id = Column(Integer, primary_key=True)
        a_id = Column(ForeignKey("a.id"))
        a = relationship("A", back_populates="bs")

    Base.metadata.create_all(e)

    s = smaker()

    a1 = A(data="x")
    s.add(a1)
    b1 = B(a=a1)

    # uncomment for program to succeed
    # s.add(b1)

    s.commit()
```

the above program intercepts relationship attribute assignments and appends.  it looks at the "right" side object for a session, and tests that the "left" is not in a session.  the "left" side then gets added to a set until it gets attached.  the commit handler makes sure nothing needing attach is left.

**Reply by CaselIT** [reply]
Maybe that example could be added to the documentation where it talks about this warning?

**Reply by zzzeek** [reply]
it's long so I'd likely add it to the attribute instrumentation examples, then we can link to it...then we can support it when it breaks hooray

**Reply by CaselIT** [reply]
sounds like a plan!

**Reply by ianjosephwilson** [reply]
This linter/sanity-checker looks good but I can't seem to get it to work with non-declarative mapped classes, is that possible ?  I'm using imperative mappings / classical mappings.  I have a base class, `ModelWithId`, for them and I try to attach the listener like this `event.listen(ModelWithId, "attribute_instrument", configure_listener)`.

**Reply by zzzeek** [reply]
My example is missing an important parameter, this should fix it:

    event.listen(Base, "attribute_instrument", configure_listener, propagate=True)

if that doesnt fix it, then there's another thing we can do to catch all the mappers but above should work

**Reply by zzzeek** [reply]
hmmmm nope I dont need propagate=True here for classical mappers, here's same program with classical mappers and a base class, no declarative:

```py
from sqlalchemy import event
from sqlalchemy.orm import object_session
from sqlalchemy.orm import RelationshipProperty


def configure_listener(class_, key, inst):
    """an 'attribute_instrument' listener.

    this is from the example at
    https://docs.sqlalchemy.org/en/14/_modules/examples/custom_attributes/listen_for_events.html

    """
    if not isinstance(inst.property, RelationshipProperty):
        return

    @event.listens_for(inst, "append")
    def append(instance, value, initiator):
        _set_needs_attach(instance, value)

    @event.listens_for(inst, "set")
    def set_(instance, value, oldvalue, initiator):
        _set_needs_attach(instance, value)

    def _set_needs_attach(instance, value):
        """act on "instance.foo = value"

        if "value" is in a session and "instance" isn't, add it to a special
        collection in that session

        """

        sess = object_session(value)
        if sess and not object_session(instance):
            if "needs_attach" not in sess.info:
                sess.info["needs_attach"] = set()
            sess.info["needs_attach"].add(instance)


def update_attached(session, instance):
    """operate on the after_attach event.  remove the instance from
    needs_attached"""

    if "needs_attach" in session.info:
        session.info["needs_attach"].discard(instance)


def verify_attached(session):
    """operate on the before_commit event.   everything should be attached."""

    if "needs_attach" in session.info:
        assert not session.info[
            "needs_attach"
        ], f"objects were unattached : {session.info['needs_attach']}"


if __name__ == "__main__":

    from sqlalchemy import (
        Column,
        Integer,
        String,
        ForeignKey,
        create_engine,
        Table,
    )
    from sqlalchemy.orm import relationship
    from sqlalchemy.orm import sessionmaker, registry

    class Base:
        pass

    e = create_engine("sqlite://")
    smaker = sessionmaker(e, future=True)

    event.listen(Base, "attribute_instrument", configure_listener)
    event.listen(smaker, "after_attach", update_attached)
    event.listen(smaker, "before_commit", verify_attached)

    reg = registry()

    class A(Base):
        pass

    reg.map_imperatively(
        A,
        Table(
            "a",
            reg.metadata,
            Column("id", Integer, primary_key=True),
            Column("data", String),
        ),
        properties={"bs": relationship("B", back_populates="a")},
    )

    class B(Base):
        a = relationship("A", back_populates="bs")

    reg.map_imperatively(
        B,
        Table(
            "b",
            reg.metadata,
            Column("id", Integer, primary_key=True),
            Column("a_id", Integer, ForeignKey("a.id")),
        ),
        properties={"a": relationship("A", back_populates="bs")},
    )
    reg.metadata.create_all(e)

    s = smaker()

    a1 = A(data="x")
    s.add(a1)
    b1 = B(a=a1)

    # uncomment for program to succeed
    # s.add(b1)

    s.commit()

```

**also**, on this whole subject, for many years now SQLAlchemy itself warns when an object in a collection is not in the session, during flush.  if you take the event listeners out of my example, you get this warning:

```
/home/classic/dev/sqlalchemy/test3.py:119: SAWarning: Object of type <B> not in session, add operation along 'A.bs' will not proceed
  s.commit()
```

I forgot all about that.  so there already is a warning for this class of error and perhaps this actually solves your problem as well.

**Reply by ianjosephwilson** [reply]
Thanks for the help.  Using this tool (the listeners) will work to migrate away from using backref cascades.

I realized after looking at your new full example the problem was that I was setting the listener after I configured the mappers.  So I think the event didn't fire.  Now it is working.  I did make a change that was running into issues when `object_session` was called on `None`.  I'm not sure if more would need to be done for non-model values that should fail a more appropriate way.

```python
# I had to alter ` _set_needs_attach` to this:
sess = object_session(value) if value is not None else None
# instead of
sess = object_session(value)
```

My tests run and I did have one occurrence of objects left unattached.  Although I use this backref cascade merging **a lot** it turns out, 298 tests and 6874 warnings.  I think a few warnings are for other things but most are for **backref cascade**.  I toggle `future` off to see the warnings and toggle it on to get the 1 failure.  I run `verify_attached` manually because my tests rollback the session after each test.  It seems that although I am triggering the cascade, by some miracle I am mostly not depending on it.

The majority of the usage involves requesting some primary object be created, and that code calling out to other functions to create secondary objects that are required to create the primary object.  Each secondary object is either already in the session or is added to the session in its own code (so it isn't left unattached) but then I go and create the primary object using those secondary objects and that triggers the exception.
     
You are right about the additional warning.  I saw it when I was testing another example, maybe I could fail on `SAWarning` and it would achieve that same affect as the above code in this case? (in addition to other common warnings)  At least in the test suite.

**Reply by zzzeek** [reply]
definitely fail on that warning, yes.   That's a warning people shouldn't get.  it's not an error because technically someone might mean to do that but it would be strange.

**Reply by zzzeek** [reply]
ideally your test suite would fail on all warnings of any kind, and setup filters only to allow the warnings that you already know about which are for the moment unavoidable.  any not-seen-before warning should fail a test suite.

**Comment by AbdealiLoKo** [comment]
Thanks for the write ups and examples @zzzeek 

I had 1 question.
With sqla 1.4 - in 1 test of mine I have ~50 warnings related to this reverse cascade.
But when I set Session.future = True - I have no warnings nor errors.

Am I right in understanding that the example:
```
d = Discussion()
session.add(d)
c = Comment(discussion=d)
session.add(c)

session.flush()
```
is throwing a warning in sqla 1.4 - but when I move to sqla 2.0 there won't be any real error ?
(Because I have ensured I did a .add() on both the objects)

It seems like sqla 1.4 is trying to tell me all the places where this error "may" occur (because in the future it is not keeping a track of the `session.add()` I may call on it)

Is my understanding right ?

**Reply by zzzeek** [reply]
Yes that's exactly right, in your code above the Comment is not added until you .add() it, whereas previously it would have been backref-cascaded.  You can use "obj in session" to test when the object is added.

**Reply by AbdealiLoKo** [reply]
Understood - thanks !
